### PR TITLE
[stdlib] Replace `precondition` with the internal `_precondition`

### PR DESCRIPTION
**Pull request by HassanElDesouky**
Replace `precondition` with the internal `_precondition` in several files in `stdlib/public/core`.

cc @lorentey, @glessard

### CODE REVIEW

**Comment by HassanElDesouky** [comment]
@lorentey There are 4 occurrences in `ArrayBuffer.swift` which still uses `precondition`. I tried to change it, but I got an error while building the stdlib. 
```
error: cannot convert value of type 'String' to expected argument type 'StaticString'
```

I tried casting the string to a static string but that didn't work. Do you have any idea why this is happening?

**Comment by lorentey** [comment]
That's because the internal assertions are (intentionally) only taking compile-time constant strings, not dynamic ones.

It looks like those ArrayBuffer preconditions serve a similar purpose as `KEY_TYPE_OF_DICTIONARY_VIOLATES_HASHABLE_REQUIREMENTS` for `Dictionary` and it probably would make sense to handle them in a similar way. (The VERY_LOUD_FUNCTION_NAME shows up in crash reports and serves as a (relatively) obvious indicator that such traps are due to a bug in the client code, not in the stdlib.)

We do want the trap message to be dynamically constructed (so that we can print the types involved) in this case, so we can't use `_precondition`. Instead, like Dictionary, these should use a regular condition and call out to the loudly named symbol in the failure case, and that ought to call `_assertionFailure`.

**Comment by lorentey** [comment]
@swift-ci test

File: stdlib/public/core/Sort.swift [hunk 1]
```diff
@@ -696,6 +696,6 @@ extension UnsafeMutableBufferPointer {
     }
 
     // FIXME: Remove this, it works around rdar://problem/45044610
```
Comment (glessard) [inline stdlib/public/core/Sort.swift:698]: This issue has perhaps been fixed, and maybe we can remove the precondition altogether.
An earlier PR to do that: https://github.com/apple/swift/pull/20714. (ping @natecook1000)
It may be worth trying to remove the precondition and then trying to run the benchmarks. If things go wrong, we can put it back.

Comment (HassanElDesouky) [reply stdlib/public/core/Sort.swift:698]: Since I don't have access to the `rdar` system, I can't really tell if it has been fixed or not. But if it's fixed I can, of course, remove the `precondition` and see how it goes.

Comment (glessard) [reply stdlib/public/core/Sort.swift:698]: Sorry for not being clear; the radar is in an in-between state, so I'm not _certain_ it's fixed. Thus my "perhaps"!  
We should just try removing it.

Comment (lorentey) [reply stdlib/public/core/Sort.swift:698]: See https://bugs.swift.org/browse/SR-14750. This worked around an optimizer issue -- it looks like the optimizer may have been confused by the created array getting immediately discarded and that triggered a miscompilation. 

The radar is just a todo item. It does not contain information that helps figuring out whether this has been fixed -- the way to do that is to try removing the workaround and see if it triggers a regression.

(Btw, we recently added a new temporary allocation facility that may be a more idiomatic way to get temporary scratch space -- cc @natecook1000)

Comment (glessard) [reply stdlib/public/core/Sort.swift:698]: ...and it if doesn't work, the rdar link should be replaced with [SR-14750](https://bugs.swift.org/browse/SR-14750).

**Review by glessard — DISMISSED** [review]

**Comment by glessard** [comment]
@swift-ci please test

File: stdlib/public/core/Sort.swift [hunk 2]
```diff
@@ -438,9 +438,6 @@ internal func _merge<Element>(
       }
     }
   }
-
-  // FIXME: Remove this, it works around rdar://problem/45044610
-  return true
```
Comment (glessard) [inline stdlib/public/core/Sort.swift:443]: If we find the workaround is no longer needed, maybe the changes to `Sort.swift` should be a separate PR!

Comment (HassanElDesouky) [reply stdlib/public/core/Sort.swift:443]: Yes sure! That make senes.

**Comment by glessard** [comment]
@swift-ci please smoke benchmark

**Comment by glessard** [comment]
[Here](https://github.com/glessard/swift/commits/stdlib-internalPrecondition) is a commit where I switched the `precondition` calls for branch-then-`_assertionFailure`.

**Comment by glessard** [comment]
The Linux test failed during testing of `swift-docc`; it has nothing to do with this.

**Comment by glessard** [comment]
@swift-ci please test linux platform

**Comment by glessard** [comment]
@swift-ci Please clean test Linux platform

**Comment by glessard** [comment]
@HassanElDesouky It _looks_ like the change to `sort` worked. I suggest that we take that particular change to a follow-up PR. The rest looks like it's in good shape.

**Comment by HassanElDesouky** [comment]
@glessard That's good news! I can cherry pick the last commit in a separate PR but I'm not free today so I'll do that tomorrow, inshallah.

**Comment by HassanElDesouky** [comment]
@glessard I cherry picked the sort optimization commit and opened a new PR for it. I also removed that commit from this branch so now this PR is only responsible for replacing `prediction` with `_prediction`

**Review by glessard — APPROVED** [review]

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@HassanElDesouky Thanks for helping with these changes!
